### ISSUE TITLE
Map renderer giving 0 size with MoveToRegion

### ISSUE THREAD

**Issue by nodoid** [xamarin/Xamarin.Forms#1852]
Bug report best practices: https://github.com/xamarin/Xamarin.Forms/wiki/Submitting-Issues

### Description
Android map renderer giving 0 size for Map
### Steps to Reproduce

1. Create a custom map
2. Move away from the map
3. Come back to the map

### Expected Behavior
The renderer displays the map

### Actual Behavior
Xamarin.Forms MapRenderer: MoveToRegion exception: Java.Lang.IllegalStateException: Error using newLatLngBounds(LatLngBounds, int): Map size can't be 0. Most likely, layout has not yet occured for the map view.  Either wait until layout has occurred or use newLatLngBounds(LatLngBounds, int, int, int) which allows you to specify the map's dimensions.
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <fcbf47a04b2e4d90beafbae627e1fca4>:0 
  at Java.Interop.JniEnvironment+InstanceMethods.CallVoidMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x00069] in <402ad96b56e643ccbd17e513ae1b6078>:0 
  at Android.Runtime.JNIEnv.CallVoidMethod (System.IntPtr jobject, System.IntPtr jmethod, Android.Runtime.JValue* parms) [0x0000e] in <183d469964214eae9915d97180f9356f>:0 
  at Android.Gms.Maps.GoogleMap.MoveCamera (Android.Gms.Maps.CameraUpdate update) [0x00048] in <bcb3af9a0115492baaa7a2e17d504257>:0 
  at Xamarin.Forms.Maps.Android.MapRenderer.MoveToRegion (Xamarin.Forms.Maps.MapSpan span, System.Boolean animate) [0x000c5] in <5930f5b120c94ff2aae31bb1351fbfd2>:0 
### Basic Information

- Version with issue: 2.5.0.122203
- Last known good version: (first reported in 2.3.4)
- IDE: VS for Mac
- Platform Target Frameworks: 
  - Android: Min - API 15, Target 26
- Android Support Library Version: 
- Nuget Packages: Xamarin.Forms-2.5.0.122203, Xam.Forms.Maps-2.5.0.122203
- Affected Devices: Android phones

**Comment by jassmith** [comment]
Do you have a reproduction or set of steps? I can't seem to cause this behavior. My maps are working fine.

**Comment by PureWeen** [comment]
If you are still having this issue please attach a repro and reopen

**Comment by carl-neng** [comment]
Hello Dear, I got the same issue above with Xamarin.Forms.Maps version 3.0.0.446417. So how to resolve it?

**Comment by nbsoftware** [comment]
I'm having this issue too... right from the app startup.

Weirdly enough, the xamarin-sample "CustomRenderer/Map" is working fine so it must something else but I fail to see what. I've spent already way too much time trying to isolate the issue, without success.

As a simple test, I modified my app to start on a single very basic page, i.e. without even moving the map:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:local="clr-namespace:Xamarin.Forms.Maps;assembly=Xamarin.Forms.Maps"
             x:Class="fipacapp.Test">
    <ContentPage.Content>
        <AbsoluteLayout VerticalOptions="Fill" HorizontalOptions="Fill">
          <local:Map
            x:Name="customMap"
            IsShowingUser="False"
            HasZoomEnabled="True"
            MapType="Street"
          AbsoluteLayout.LayoutFlags="All"
          AbsoluteLayout.LayoutBounds="0,0,1,.5" />
        </AbsoluteLayout>
    </ContentPage.Content>
</ContentPage>
```

And I get this exception below. Does it help someone here to have a clue of what is going on here?

As for some background info, my app was working fine when targeting Android 8.
This all started when trying to upgrade to AndroidX, which I was very close to, if it wasn't for this annoying issue I can't seem to resolve.

```
...
[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'com/android/okhttp/internal/huc/HttpURLConnectionImpl'
[Mono] Requesting loading reference 3 (of 5) of /storage/emulated/0/Android/data/com.fipac.android.fipacapp/files/.__override__/Xamarin.GooglePlayServices.Maps.dll
[Mono] Loading reference 3 of /storage/emulated/0/Android/data/com.fipac.android.fipacapp/files/.__override__/Xamarin.GooglePlayServices.Maps.dll asmctx DEFAULT, looking for Xamarin.GooglePlayServices.Basement, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
[Mono] Assembly Ref addref Xamarin.GooglePlayServices.Maps[0x7c12187b0480] -> Xamarin.GooglePlayServices.Basement[0x7c12187a4c80]: 3
[Mono] Requesting loading reference 9 (of 10) of /storage/emulated/0/Android/data/com.fipac.android.fipacapp/files/.__override__/Xamarin.Forms.Maps.Android.dll
[Mono] Loading reference 9 of /storage/emulated/0/Android/data/com.fipac.android.fipacapp/files/.__override__/Xamarin.Forms.Maps.Android.dll asmctx DEFAULT, looking for Xamarin.AndroidX.Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
[Mono] Assembly Ref addref Xamarin.Forms.Maps.Android[0x7c121878c500] -> Xamarin.AndroidX.Core[0x7c1218682600]: 8
[AppCenter] checkPendingLogs(group_core) pendingLogCount=0 batchTimeInterval=3000
[AppCenter] Calling https://in.appcenter.ms/logs?api-version=1.0.0...
[AppCenter] Headers: {Install-ID=f2899f44-bf2d-4832-8f63-13c4968f0057, App-Secret=****************************d83ef9eb, Content-Type=application/json}
[AppCenter] {
[AppCenter]   "logs": [
[AppCenter]     {
[AppCenter]       "type": "startService",
[AppCenter]       "timestamp": "2020-05-29T08:31:07.822Z",
[AppCenter]       "device": {
[AppCenter]         "wrapperSdkVersion": "3.2.1",
[AppCenter]     05-29 10:31:11.266 V/AppCenter( 2609):         "wrapperSdkName": "appcenter.xamarin",
[AppCenter]         "wrapperRuntimeVersion": "10.3.1.0;",
[AppCenter]         "sdkVersion": "3.2.1",
[AppCenter]         "model": "Android SDK built for x86_64",
[AppCenter]         "oemName": "Google",
[AppCenter]         "osName": "Android",
[AppCenter]         "osVersion": "10",
[AppCenter]         "osBuild": "QSR1.200403.001",
[AppCenter]         "appBuild": "1905-29 10:31:11.266 V/AppCenter( 2609):         "osApiLevel": 29,
[AppCenter]         "locale": "en_US",
[AppCenter]         "timeZoneOffset": 120,
[AppCenter]         "screenSize": "1080x2088",
[AppCenter]         "appVersion": "1.9.0",
[AppCenter]         "carrierName": "Android",
[AppCenter]         "carrierCountry": "us",
[AppCenter]         "appBuild": "190001",
[AppCenter]         "appNamespace": "com.fipac.android.fipacapp"
[AppCenter]       },
[AppCenter]       "services": [
[AppCenter]         "Analytics",
[AppCenter]         "Crashes"
[AppCenter]       ]
[AppCenter]     }
[AppCenter]   ]
[AppCenter] }
[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'com/android/okhttp/okio/RealBufferedSource$1'
[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'java/util/Collections$UnmodifiableCollection$1'
[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'com/google/maps/api/android/lib6/common/apiexception/c'
[Mono] DllImport searching in: '__Internal' ('(null)').
[Mono] Searching for 'java_interop_jnienv_call_boolean_method'.
[Mono] Probing 'java_interop_jnienv_call_boolean_method'.
[Mono] Found as 'java_interop_jnienv_call_boolean_method'.
[Mono] DllImport searching in: '__Internal' ('(null)').
[Mono] Searching for 'java_interop_jnienv_new_local_ref'.
[Mono] Probing 'java_interop_jnienv_new_local_ref'.
[Mono] Found as 'java_interop_jnienv_new_local_ref'.
Xamarin.Forms MapRenderer: MoveToRegion exception: Java.Lang.IllegalStateException: Error using newLatLngBounds(LatLngBounds, int): Map size can't be 0. Most likely, layout has not yet occured for the map view.  Either wait until layout has occurred or use newLatLngBounds(LatLngBounds, int, int, int) which allows you to specify the map's dimensions.
  at Java.Interop.JniEnvironment+InstanceMethods.CallNonvirtualVoidMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniObjectReference type, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args) [0x0008e] in <9324da45a6654f83baffa7c2854d836a>:0 
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeNonvirtualVoidMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x0001f] in <9324da45a6654f83baffa7c2854d836a>:0 
  at Android.Gms.Maps.GoogleMap.MoveCamera (Android.Gms.Maps.CameraUpdate update) [0x00027] in D:\a\1\s\generated\com.google.android.gms.play-services-maps\obj\Release\monoandroid90\generated\src\Android.Gms.Maps.GoogleMap.cs:3816 
  at Xamarin.Forms.Maps.Android.MapRenderer.MoveToRegion (Xamarin.Forms.Maps.MapSpan span, System.Boolean animate) [0x000c5] in <47a2a91c56584498b8fe9e2ebc17ae05>:0 
  --- End of managed Java.Lang.IllegalStateException stack trace ---
com.google.maps.api.android.lib6.common.apiexception.c: Error using newLatLngBounds(LatLngBounds, int): Map size can't be 0. Most likely, layout has not yet occured for the map view.  Either wait until layout has occurred or use newLatLngBounds(LatLngBounds, int, int, int) which allows you to specify the map's dimensions.
	at com.google.maps.api.android.lib6.common.m.b(:com.google.android.gms.dynamite_mapsdynamite@200414056@20.04.14 (040800-0):5)
	at com.google.maps.api.android.lib6.gmm6.api.c.a(:com.google.android.gms.dynamite_mapsdynamite@200414056@20.04.14 (040800-0):39)
	at com.google.maps.api.android.lib6.impl.y.a(:com.google.android.gms.dynamite_mapsdynamite@200414056@20.04.14 (040800-0):0)
	at com.google.maps.api.android.lib6.gmm6.api.c.a(:com.google.android.gms.dynamite_mapsdynamite@200414056@20.04.14 (040800-0):48)
	at com.google.maps.api.android.lib6.impl.bi.a(:com.google.android.gms.dynamite_mapsdynamite@200414056@20.04.14 (040800-0):250)
	at com.google.android.gms.maps.internal.i.a(:com.google.android.gms.dynamite_mapsdynamite@200414056@20.04.14 (040800-0):175)
	at ch.onTransact(:com.google.android.gms.dynamite_mapsdynamite@200414056@20.04.14 (040800-0):4)
	at android.os.Binder.transact(Binder.java:914)
	at com.google.android.gms.internal.maps.zza.zzb(Unknown Source:20)
	at com.google.android.gms.maps.internal.zzg.moveCamera(Unknown Source:20)
	at com.google.android.gms.maps.GoogleMap.moveCamera(Unknown Source:13)
	at crc648aad9efe354a1d8f.MapRenderer.n_onLayout(Native Method)
	at crc648aad9efe354a1d8f.MapRenderer.onLayout(MapRenderer.java:49)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at com.xamarin.forms.platform.android.FormsViewGroup.measureAndLayout(FormsViewGroup.java:37)
	at crc643f46942d9dd1fff9.Platform_DefaultRenderer.n_onLayout(Native Method)
	at crc643f46942d9dd1fff9.Platform_DefaultRenderer.onLayout(Platform_DefaultRenderer.java:72)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at com.xamarin.forms.platform.android.FormsViewGroup.measureAndLayout(FormsViewGroup.java:37)
	at crc643f46942d9dd1fff9.PageRenderer.n_onLayout(Native Method)
	at crc643f46942d9dd1fff9.PageRenderer.onLayout(PageRenderer.java:72)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at com.xamarin.forms.platform.android.FormsViewGroup.measureAndLayout(FormsViewGroup.java:37)
	at crc643f46942d9dd1fff9.PlatformRenderer.n_onLayout(Native Method)
	at crc643f46942d9dd1fff9.PlatformRenderer.onLayout(PlatformRenderer.java:63)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1103)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
	at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
	at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
	at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
	at com.android.internal.policy.DecorView.onLayout(DecorView.java:779)
	at android.view.View.layout(View.java:21912)
	at android.view.ViewGroup.layout(ViewGroup.java:6260)
	at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:3080)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2590)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1721)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7598)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:966)
	at android.view.Choreographer.doCallbacks(Choreographer.java:790)
	at android.view.Choreographer.doFrame(Choreographer.java:725)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:951)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7356)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)

[monodroid-assembly] typemap: unable to find mapping to a managed type from Java type 'java/util/Collections$UnmodifiableCollection$1'
...
```

**Comment by AceCoderLaura** [comment]
I'm getting this problem and my XAML is nearly identical to the example [here](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/map/pins#display-a-pin-collection).

Here's mine:
```
<?xml version="1.0" encoding="utf-8"?>

<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:xfs="clr-namespace:Xamarin.Forms;assembly=Xamarin.Forms.Core"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:local="clr-namespace:Xamarin.Scratch"
             xmlns:maps="clr-namespace:Xamarin.Forms.Maps;assembly=Xamarin.Forms.Maps"
             xmlns:taskManagement="clr-namespace:Adapt.Model.TaskManagement;assembly=Adapt.Model.Common.Standard"
             x:Class="Xamarin.Scratch.MainPage">

    <ContentPage.Resources>
        <local:TaskList x:Key="TaskList" />
        <local:PointItemToMapPinConverter x:Key="PointItemToMapPinConverter" />
    </ContentPage.Resources>

    <Grid>
        <maps:Map ItemsSource="{StaticResource TaskList}">
            <maps:Map.ItemTemplate>
                <DataTemplate x:DataType="{x:Type taskManagement:TaskInfo}">
                    <maps:Pin Position="{Binding PointItem, Converter={x:StaticResource PointItemToMapPinConverter}}"
                              Address="{Binding Address.AddressDescription}"
                              Label="{Binding TaskNo}" />
                </DataTemplate>
            </maps:Map.ItemTemplate>
        </maps:Map>
    </Grid>

</ContentPage>
```

**Comment by nbsoftware** [comment]
Did anyone find at least a work around?

This thing is driving me crazy and is preventing me from delivering an Android version to my customer, without me being able to tell when I'll be able to. Ouch.

**Comment by AceCoderLaura** [comment]
@nbsoftware I don't know what I did to fix the issue but I think it might have had something to do with my API key. I spent a lot of time referencing [this article](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/map/setup#android) and messing around in the manifest. Once it was working I tried to work backwards to find what made it start working again but even when I reset my branch back to master it kept working. So my best guess is that it's something to do with API keys somehow getting "stale" and using a new one seems to "refresh" the old one?

Sorry it's not much help but that's all I could deduce.

TL;DR:
- Try a new API key
- Try cleaning your repository and building from scratch

**Comment by nbsoftware** [comment]
Hello,

Thank you for you advice.

I ended up rebuilding the whole Android project from scratch and targeting AndroidX (Q) right from the start. It took me a hell lot of time to do that but it was worth it: it works now as expected (no more Map initialisation errors) and the whole project is a lot cleaner with a lot less dependencies. Worth pointing, I started a new project with Visual Studio 2019 on Windows to get the "reference" project style rather than using package.config file. Then reopened that skeleton project on Mac.

I guess I'll never know what went wrong, but I suppose the AndroidX migration step that VS proposes might be the culprit.
I find the whole Xamarin / Forms ecosystem very very fragile and unstable. 

Anyway, moving on right now: on to next bug...

I will never
